File: Buzz/Core/Models/Organization.swift
import SwiftUI

struct Organization: Identifiable, Codable, Hashable, Sendable {
    let id: UUID
    var name: String
    var handle: String                  // e.g. "isa-ucsd"
    var tagline: String
    var description: String
    var category: OrganizationCategory
    var campus: String                  // e.g. "ucsd"
    var foundedYear: Int?
    var memberCount: Int
    var logoURL: URL?
    var coverURL: URL?
    var accentHex: String               // hex like "#FF2D92" — drives the org's visual identity
    var isVerified: Bool
    var instagramHandle: String?        // bare handle, no "@" — e.g. "acm.ucsd"
    var websiteURL: URL?                // org's canonical site

    var accent: Color { Color(hex: accentHex) ?? .accentColor }

    // Canonicalised Instagram URL. Nil when we don't trust the handle (prevents phishing via
    // raw user-entered "insta.sketchy/login" style values sneaking through).
    var instagramURL: URL? {
        guard let raw = instagramHandle?.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.isEmpty else { return nil }
        let cleaned = raw.hasPrefix("@") ? String(raw.dropFirst()) : raw
        // Instagram allows letters, numbers, periods, underscores. Reject anything else.
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._"))
        guard !cleaned.isEmpty, cleaned.unicodeScalars.allSatisfy({ allowed.contains($0) }) else {
            return nil
        }
        return URL(string: "https://instagram.com/\(cleaned)")
    }

    // Only surface https websites — a club link shouldn't deep-link into sketchy schemes.
    var safeWebsiteURL: URL? {
        guard let url = websiteURL, let scheme = url.scheme?.lowercased(),
              scheme == "https" || scheme == "http" else { return nil }
        return url
    }
}

// Tiny hex helper colocated since it's only used by Organization right now.
extension Color {
    init?(hex: String) {
        var s = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6, let v = UInt64(s, radix: 16) else { return nil }
        self = Color(
            red:   Double((v >> 16) & 0xFF) / 255.0,
            green: Double((v >>  8) & 0xFF) / 255.0,
            blue:  Double( v        & 0xFF) / 255.0
        )
    }
}

File: android/app/src/main/kotlin/com/arnavgoel/buzz/data/model/Organization.kt
package com.arnavgoel.buzz.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Organization(
    val id: String,
    val name: String,
    val handle: String,
    val tagline: String = "",
    val description: String = "",
    val category: OrganizationCategory = OrganizationCategory.GENERAL,
    val campus: String,
    @SerialName("founded_year") val foundedYear: Int? = null,
    @SerialName("member_count") val memberCount: Int = 0,
    @SerialName("logo_url") val logoUrl: String? = null,
    @SerialName("cover_url") val coverUrl: String? = null,
    @SerialName("accent_hex") val accentHex: String = "#C850FF",
    @SerialName("is_verified") val isVerified: Boolean = false,
    @SerialName("instagram_handle") val instagramHandle: String? = null,
    @SerialName("website_url") val websiteUrl: String? = null
) {
    /**
     * Mirrors iOS `Organization.instagramURL`. Only accepts the Instagram-allowed
     * character set (alphanumerics + . + _) so a raw user-entered "insta.sketchy/login"
     * style value can't sneak through into a tappable link.
     */
    val safeInstagramUrl: String?
        get() {
            val raw = instagramHandle?.trim().orEmpty()
            if (raw.isEmpty()) return null
            val cleaned = if (raw.startsWith("@")) raw.drop(1) else raw
            if (cleaned.isEmpty()) return null
            val allowed = cleaned.all { it.isLetterOrDigit() || it == '.' || it == '_' }
            return if (allowed) "https://instagram.com/$cleaned" else null
        }

    /** Mirrors iOS `safeWebsiteURL` — only http/https survive. */
    val safeWebsiteUrl: String?
        get() {
            val url = websiteUrl ?: return null
            val lower = url.lowercase()
            return if (lower.startsWith("https://") || lower.startsWith("http://")) url else null
        }
}

@Serializable
enum class OrganizationCategory(val displayName: String) {
    @SerialName("greek")        GREEK("Greek life"),
    @SerialName("academic")     ACADEMIC("Academic"),
    @SerialName("cultural")     CULTURAL("Cultural"),
    @SerialName("religious")    RELIGIOUS("Religious"),
    @SerialName("sports")       SPORTS("Sports"),
    @SerialName("arts")         ARTS("Arts"),
    @SerialName("service")      SERVICE("Service"),
    @SerialName("professional") PROFESSIONAL("Professional"),
    @SerialName("political")    POLITICAL("Political"),
    @SerialName("identity")     IDENTITY("Identity"),
    @SerialName("general")      GENERAL("General")
}
